### ISSUE TITLE
_sync:seq less than document sequence values causing missing _changes

### ISSUE THREAD

**Issue by adamcfraser** [couchbase/sync_gateway#1090]
User reported the _changes feed not returning recent documents for a channel.  Based on the information provided, the documents appear to be getting channeled properly.  Restarting Sync Gateway doesn't change anything, so this doesn't look like a case of Sync Gateway dropping the server mutation feed.

In particular, it looks like these docs aren't being returned by the view-based channel cache backfill.  Using this ticket to track followup with user to understand whether this is due to the document not being indexed, or the range being requested by Sync Gateway.

**Comment by adamcfraser** [comment]
Background: 
https://forums.couchbase.com/t/sync-gateway--changes-feed-and-pull-replication-does-not-return-all-documents-up-to-the-latest-sequence/4918

**Comment by alex7egli** [comment]
I ran a _resync a few days ago, not sure if that could cause this issue but just in case it is relevant. Especially since the sequence numbers seem off. The documents in the channel being pulled that are not being returned have a higher sequence number than the last_seq being returned by the _changes call, and any documents with a lower sequence number than the last_seq returned are showing up in the Pull replication and _changes request.

**Comment by alex7egli** [comment]
My issue seems very similar to this other issue that was closed due to inactivity: https://github.com/couchbase/couchbase-lite-ios/issues/608

**Comment by adamcfraser** [comment]
I agree there are some similarities, but a few notable differences:
1. Based on the sample doc you provided, the document has the appropriate channels assigned to it.
2. You're still seeing the problem when running with since=0, so it's not the case that the client has a since value higher than the missing docs.

A few more questions, to rule out some possibilities:
1. Are the missing documents particularly large (i.e. larger than 1MB)?  There was an issue in older versions of Couchbase Server where large docs weren't being indexed properly.
2. What's the value of the _sync:seq document in your database?

**Comment by alex7egli** [comment]
1. No, they are json documents. One of them is 17KB, and they are all around the same size, not varying by more than a few KB. The one that is 17KB has "sequence": 2812497
2. _sync:seq = 2762344.

As a temporary fix, is there any sort of command I can run on sync gateway to get it to fix the sequence number and update it to the latest value? If I manually change the value of it on the server in the _sync:seq document to a much higher value to make sure it returns everything would that break anything?

**Comment by adamcfraser** [comment]
The value of _sync:seq is unexpected.  There's shouldn't be a scenario where there are documents with sequence values higher than _sync:seq in your database, as _sync:seq is what's used to generate the sequence numbers in the first place. 

I don't know how that could happen.  Was there any backup/restore of your couchbase server that could have rolled back the _sync:seq value?

**Comment by adamcfraser** [comment]
I'll look into whether there's anything happening during _resync that could cause this.  It seems unlikely, though, as Sync Gateway only ever issues Increment operations against _sync:seq - it doesn't try to set the value directly.

**Comment by alex7egli** [comment]
Yes, this is a copy of our production database. I ran a cbrestore using a cbbackup of our prod database last week. I'm just re-running the cbrestore --add operation again to see if I missed one of our backups (we have two nodes so I have to run the backup operation against both and then run the cbrestore --add operation with both backups on our test server). I checked the _sync:seq doc of the db this is a copy of and it is much higher, though that was a few days ago so a lot of changes have happened on that other database.

**Comment by adamcfraser** [comment]
The root cause of your _changes issues is definitely the incorrect _sync:seq value.  I'll wait to see whether you can sort out whether this was an issue of the backup/restore not being complete.

**Comment by alex7egli** [comment]
I just re-ran the cbrestore --add command and the _sync:seq value was not updated, which means either the cbrestore command failed to update the document, or the original cbbackup failed to backup the document. If I manually update _sync:seq to the value it is on our other server (which will be higher than what the actual value should be) would it cause any problems?

**Comment by adamcfraser** [comment]
Yes, there's the potential for other problems if you've got inconsistency between _sync:seq and the document sequences in your database (even if it's set to a higher value).  You'll probably need to get to the bottom of why your backup/restore doesn't represent a valid snapshot of your production environment.

**Comment by alex7egli** [comment]
The cbbackup command I used was:
sudo /opt/couchbase/bin/cbbackup couchstore-files:///couchbase-data /datadrive/db_backup -u `<couchbase username>` -p `<couchbase password>`

The cbrestore command I used was:
sudo /opt/couchbase/bin/cbrestore -b <bucket name> --add /datadrive/db_backup http://<hostname>:8091 -u `<couchbase username>` -p `<couchbase password>`

I ran both commands once for each server in our cluster. I used the couchstore-files restore since running the normal restore command against a running cluster will use up all RAM and crash, but the files restore runs ok and doesn't affect the performance of the server it's being run on. If one or both of these commands doesn't work it's very troubling for us, since that is the only way we can backup our production data.

**Comment by adamcfraser** [comment]
If you're running the backups independently for each server in the cluster, while that server is still live (i.e. new documents are being written), then I think you're likely to end up with an inconsistent overall cluster when you do your restore.
e.g. If _sync:seq is stored in node 1:
1. Backup node 1, get _sync:seq
2. Several documents get written to node 2, which would also have incremented _sync:seq
3. Backup node 2 - you'll get the new documents, but not the new _sync:seq/

**Comment by alex7egli** [comment]
Does that mean there's no way to do a proper backup of a couchbase cluster? How would we get a safe backup in case something goes wrong and we lose a node or the whole cluster?

**Comment by jchris** [comment]
Hi @alex7egli the issue is that even with a perfect backup restore, if the Couchbase Server is taking writes during backup, there can be some writes that happen after the backup finishes on node1, but before it finishes on node2. If you have documents that reference other documents, this shows up in applications as references to missing documents or old versions of documents.

There are some thing we can do to mitigate that case, and we are talking to the cbrestore team before we give advice b/c maybe they have something already and it's just a matter of getting you a preview build of the tool.

Luckily we think that if you manually set the _sync:seq to some value higher than the max value that had previously been written, the worst thing that can happen is that it masks the problem of some documents being written after backup completed. Assuming your backup/restore was complete and correct (on each individual node) then bumping that value up will make SG pick up from the new sequence and return to normal operation. Please test and verify before you erase any backup files.

We'll let you know as we learn more about options for ensuring backups choose the same checkpoint for completion across the cluster. If you end up with lingering questions about other data, we may want to have a look over your shoulder at the cluster, if that is possible.

**Comment by alex7egli** [comment]
I deleted the _sync:seq document on our test server, then re-ran the cbrestore command and this time it created the document with a much higher value. So it looks like the --add command will only add new documents, and not update existing ones. I didn't know that was the functionality of it before. It will probably take longer but I'll leave out the --add flag in the future.

Thanks for the info about cbrestore. I think for now I will settle for just scheduling the backup to run at a time when we have very little load on the server and hopefully that will mean that we only have a few documents ahead of the sync value. I can also try the manual bump up you suggested if we keep hitting this problem in the future and see if sync gateway handles it ok.

**Comment by zgramana** [comment]
@JeffThomasWrite Please add a docs portal ticket to provide basic information on using cb-backup with SG.

**Comment by JeffThomasWriter** [comment]
I entered issue 253 in the mobile portal repository. But is this CBS documentation. Also: from the above, I do not have the info I need to document.

**Comment by push-build-user** [comment]
The problem was that we had to use the couchstore-files backup method because of a bug in Couchbase Server 3.0.x cbbackup where it uses up all memory if run against the cluster. Using the couchstore-files method meant we had to backup the database files on each node individually, so the _sync:seq document gets backed up while changes are still being made to the server and it ends up being less than the actual sequence value. On the new server sync gateway can't handle this and just keeps re-indexing its views and nothing can be queried.
We solved the issue by writing a view that would find the real sequence value for the current database. After restoring the backup to a new cluster we would run the view first, then update the _sync:seq document manually with this value, then start up sync gateway. This works for us and sync gateway functions normally afterward.
I don't know if the cbbackup against the cluster will work properly either, since there is always a chance it will backup the _sync:seq document and then changes are written to other documents that it backs up after. Until we upgrade to couchbase 3.1.0 we won't be able to test cbbackup against a cluster with sync gateway.

**Comment by alex7egli** [comment]
The problem was that we had to use the couchstore-files backup method because of a bug in Couchbase Server 3.0.x cbbackup where it uses up all memory if run against the cluster. Using the couchstore-files method meant we had to backup the database files on each node individually, so the _sync:seq document gets backed up while changes are still being made to the server and it ends up being less than the actual sequence value. On the new server sync gateway can't handle this and just keeps re-indexing its views and nothing can be queried.
We solved the issue by writing a view that would find the real sequence value for the current database. After restoring the backup to a new cluster we would run the view first, then update the _sync:seq document manually with this value, then start up sync gateway. This works for us and sync gateway functions normally afterward.
I don't know if the cbbackup against the cluster will work properly either, since there is always a chance it will backup the _sync:seq document and then changes are written to other documents that it backs up after. Until we upgrade to couchbase 3.1.0 we won't be able to test cbbackup against a cluster with sync gateway.
